src/translation.h

```diff
@@ -0,0 +1,283 @@
+	}
+	class DBString;
+}
+
+
+/**
+ * Namespace used to retrieve translations of Terms, BattlEvents, etc.
+ */
+namespace Tr {
+	/**
+	 * The name of the translation directory.
+	 * @return The translation directory.
+	 */
+	std::string TranslationDir();
+
+	/**
+	 * The id of the current translation (e.g., "Spanish"). If empty, there is no active translation.
+	 * @return The translation ID
+	 */
+	std::string CurrTranslationId();
+
+} // End namespace Tr
+
+
+//////////////////////////////////////////////////////////
+// NOTE: The code for Entry and Dictionary is duplicated in LcfTrans.
+//       At some point it should be merged to a common location.
+//////////////////////////////////////////////////////////
+
+/**
+ * An entry in the dictionary
+ */
+class Entry {
+public:
+	std::string original; // msgid
+	std::string translation; // msgstr
+	std::string context; // msgctxt
+};
+
+/**
+ * A .po file loaded into memory. Contains a dictionary of entries.
+ */
+class Dictionary {
+public:
+	/**
+	 * Super simple parser.
+	 * Only parses msgstr, msgid and msgctx
+	 * Returns false on failure.
+	 *
+	 * @param res The dictionary to store the translated entries in.
+	 * @param in The stream to load the translated entries from.
+	 * @param return True if the file was loaded without error; false otherwise.
+	 */
+	static bool FromPo(Dictionary& res, std::istream& in);
+
+	/**
+	 * Replace an original string with the translated string.
+	 * Template can be "std::string" or "lcf::DBString"
+	 *
+	 * @param context The 'context' of this string; used to differentiate strings that have the same
+	 *                original language value but different target language values.
+	 * @param original The string to lookup. Will be replaced if a lookup is found.
+	 * @return True if the original string was replaced; false otherwise.
+	 */
+	template <class StringType>
+	bool TranslateString(const std::string& context, StringType& original) const;
+
+private:
+	/**
+	 * Add an entry to the dictionary.
+	 *
+	 * @param entry The entry to add.
+	 */
+	void addEntry(const Entry& entry);
+
+	// Lookup by context, where context can be empty ("") for no context.
+	std::unordered_map<std::string, std::unordered_map<std::string, std::string>> entries;
+};
+
+
+// Template implementation
+template <class StringType>
+bool Dictionary::TranslateString(const std::string& context, StringType& original) const
+{
+	std::stringstream key;
+	key <<original;
+	auto it = entries.find(context);
+	if (it != entries.end()) {
+		auto it2 = it->second.find(key.str());
+		if (it2 != it->second.end()) {
+			original = StringType(it2->second);
+			return true;
+		}
+	}
+	return false;
+}
+
+
+/**
+ * Properties of a language
+ */
+struct Language {
+	std::string langDir;  // Language directory (e.g., "en", "English_Localization")
+	std::string langName; // Display name for this language (e.g., "English")
+	std::string langDesc; // Helper text to show when the menu is highlighted
```

Comment:
should be ``lang_dir, lang_name, lang_desc``

Reply:
Fixed